<hr style="border-width:2px;border-color:#75DFC1">
<h1 style = "text-align:center" > Statistiques exploratoires </h1> 
<h2 style = "text-align:center"> Analyse descriptive des variables numériques d'un jeu de données </h2> 
<hr style="border-width:2px;border-color:#75DFC1">

### Contexte et objectif

> En datascience, avant de commencer le processing et la modélisation des données, il est très important de se familiariser avec le jeu de données dont nous disposons. Pour cela, l'analyse descriptive des variables numériques au travers d'indicateurs simples apparaît comme essentielle et permet de faire une première bonne exploration. On peut distinguer deux familles : les indicateurs de position et les indicateurs de dispersion. **Pandas**, à travers sa méthode **describe**, offre un outil permettant d'obtenir rapidement les statistiques descriptives de variables quantitatives.

> L'objectif de ce module est de comprendre quelques statistiques présentées dans l'outil **describe** et d'aller un peu plus loin dans leur interprétation et leur utilisation.

> Commençons par la phase d'importation des packages.

* Importer les packages **pandas** et **numpy**.

* Charger dans un tableau nommé *data* les données situées dans le fichier *crowdfunding.csv* en mettant la **première colonne en indice** et afficher les 5 premières lignes.

In [15]:
# Insérez votre code ici






                                                         name        category  \
ID                                                                              
1000002330                    The Songs of Adelaide & Abullah          Poetry   
1000003930      Greeting From Earth: ZGAC Arts Capsule For ET  Narrative Film   
1000004038                                     Where is Hank?  Narrative Film   
1000007540  ToshiCapital Rekordz Needs Help to Complete Album           Music   
1000011046  Community Film Project: The Art of Neighborhoo...    Film & Video   

           main_category currency    deadline     goal             launched  \
ID                                                                            
1000002330    Publishing      GBP  2015-10-09   1000.0  2015-08-11 12:12:28   
1000003930  Film & Video      USD  2017-11-01  30000.0  2017-09-02 04:43:57   
1000004038  Film & Video      USD  2013-02-26  45000.0  2013-01-12 00:20:50   
1000007540         Music      USD  20

* Déterminer les variables numériques et les stocker dans un tableau **`num_data`**

<div class="alert alert-info">
<i class="fa fa-info-circle"></i> &emsp; 
On pourra s'aider de la fonction pandas.DataFrame.select_dtypes() et de l'attribut include.
</div>


In [16]:
# Insérez votre code ici





,goal,pledged,backers
ID,,,
1000002330,1000.0,0.0,0
1000003930,30000.0,2421.0,15
1000004038,45000.0,220.0,3
1000007540,5000.0,1.0,1
1000011046,19500.0,1283.0,14


> Il existe certaines statistiques descriptives simples à calculer qui permettent d'obtenir une meilleure compréhension des données quantitatives. C'est ce qu'on appelle les indicateurs de position. Ils sont au nombre de cinq : moyenne, médiane, quantiles, min, max. La moyenne permet de résumer facilement une liste de données numériques en une seule valeur. Elle se calcule comme suit :

><center>   
$$ \overline{X} = \dfrac{1}{N}\displaystyle\sum\limits_{i=1}^Nx_i $$
</center>

> Où $\overline{X}$ est la moyenne, N la longueur de la liste et $x_i$ les éléments de la liste.


* Calculer la moyenne de chaque variable contenue dans *num_data* et les stocker dans un DataFrame **`stats`** avec comme unique colonne *moyenne*. 
* Comparer la *moyenne* des collectes (**pledged**) avec celle des objectifs fixés (**goal**). Que constatez-vous ?

> <div class="alert alert-info">
<i class="fa fa-info-circle"></i> &emsp; 
Pour la création du DataFrame stats, on pourra utiliser pd.DataFrame et l'attribut columns.
</div>

> <div class="alert alert-info">
<i class="fa fa-info-circle"></i> &emsp; 
Pour le calcul des moyennes, on pourra s'aider de pandas.DataFrame.mean().
</div>


In [17]:
# Insérez votre code ici







              moyenne
goal     39327.861938
pledged  11267.140240
backers    110.666500 



> La moyenne est une bonne entrée en matière, cependant elle ne permet pas d'avoir une répartition précise des données. Pour cela, la médiane est un indicateur plus précis. Elle se définit ainsi : "la médiane d'un ensemble de valeurs est la valeur x qui permet de couper l'ensemble des valeurs en deux parties égales : mettant d'un côté une moitié des valeurs, qui sont toutes inférieures ou égales à x et de l'autre côté l'autre moitié des valeurs, qui sont toutes supérieures ou égales à x (s'il y a un nombre impair de valeurs, la valeur centrale sera mise des deux côtés)".

* Calculer la **médiane** de chaque variable contenue dans num_data et ajouter une colonne **mediane** à *stats*.
* Interpréter ces valeurs pour chaque variable en vous servant de la définition donnée précédemment.

<div class="alert alert-info">
<i class="fa fa-info-circle"></i> &emsp; 
Pour le calcul des médianes, on pourra s'aider de pandas.DataFrame.median().
</div>

In [18]:
# Insérez votre code ici






              moyenne  mediane
goal     39327.861938   5000.0
pledged  11267.140240    650.0
backers    110.666500     12.0


> Malgré une moyenne élevée, on constate que la moitié des objectifs fixés ne dépassent pas 5000€.
> De même pour les dons récoltés et le nombre de contributeurs qui ont respectivement la moitié en dessous de 650€ et 12 personnes.
> Cela laisse apparaître la présence de valeurs extrêmes qui tirent vers le haut la moyenne.
> On observe déjà la faiblesse d'un indicateur comme la moyenne qui est fortement influencée par les valeurs extrêmes.

* Créer une variable **mean_med_diff** correspondant à la valeur absolue de la différence entre **moyenne** et **médiane**.

In [20]:
# Insérez votre code ici






,moyenne,mediane,median,mean_med_diff
goal,39327.861938,5000.0,5000.0,34327.861938
pledged,11267.140240,650.0,650.0,10617.140240
backers,110.666500,12.0,12.0,98.666500


> Pour donner une interprétation de cette nouvelle variable, il faut bien avoir en tête les ordres de grandeur de chaque variable. Dans notre cas, on constate à travers la variable nouvellement créée que la médiane et la moyenne sont systématiquement très éloignées. Cela nous renseigne sur **l'hétérogénéité** des variables dont nous disposons. Il semble clairement qu'on dispose d'un nombre important de campagne de financement avec de petits montants et d'un nombre moindre de campagne avec de gros montants. 

> Pour confirmer notre analyse, il est nécessaire de regarder les **quantiles** de chaque variable.

> Les **quantiles** sont les valeurs qui divisent un jeu de données en intervalles contenant le même nombre de données. De manière générale, on divise le jeu de données en 4 de telle sorte que le premier quantile renvoie la valeur à partir de laquelle on retrouve 25% des valeurs et ainsi de suite pour les autres quantiles.

* Déterminer les **quantiles** de goal.

<div class="alert alert-info">
<i class="fa fa-info-circle"></i> &emsp; 
Pour déterminer les quantiles on pourra s'aider de pandas.DataFrame.quantile(q = [0.25,0.5,0.75]).
</div>


In [21]:
# Insérez votre code ici






0.25     2000.0
0.50     5000.0
0.75    17500.0
Name: goal, dtype: float64

* Ajouter au tableau *stats* 3 colonnes q1,q2 et q3 renseignant respectivement le premier, deuxième et troisième quantile de chaque variable numérique.

In [23]:
# Insérez votre code ici






              moyenne  mediane  median  mean_med_diff      q1      q2       q3
goal     39327.861938   5000.0  5000.0   34327.861938  2000.0  5000.0  17500.0
pledged  11267.140240    650.0   650.0   10617.140240    33.0   650.0   4233.0
backers    110.666500     12.0    12.0      98.666500     2.0    12.0     56.0


* Afficher *stats*, comparer **q2** et **q3**.
* Comparer **q3** et la moyenne.
* Cela confirme-t-il notre analyse précédente ?

In [25]:
# Insérez votre code ici







              moyenne  mediane  median  mean_med_diff      q1      q2       q3
goal     39327.861938   5000.0  5000.0   34327.861938  2000.0  5000.0  17500.0
pledged  11267.140240    650.0   650.0   10617.140240    33.0   650.0   4233.0
backers    110.666500     12.0    12.0      98.666500     2.0    12.0     56.0

 goal       12500.0
pledged     3583.0
backers       44.0
dtype: float64

 goal       21827.861938
pledged     7034.140240
backers       54.666500
dtype: float64


> On peut compléter notre analyse en s'intéressant au min, max de chaque variable. La différence des deux nous donnera une idée de l'étendue sur laquelle se répartissent les valeurs.

* Ajouter au tableau *stats* 3 colonnes **min**,**max** et **min_max_diff** renseignant le min,le max et la différence des deux.
* Que remarquez-vous sur la différence des deux ?

In [27]:
# Insérez votre code ici





              moyenne  mediane  median  mean_med_diff      q1      q2  \
goal     39327.861938   5000.0  5000.0   34327.861938  2000.0  5000.0   
pledged  11267.140240    650.0   650.0   10617.140240    33.0   650.0   
backers    110.666500     12.0    12.0      98.666500     2.0    12.0   

              q3  min          max  min_max_diff  
goal     17500.0  1.0  40000000.00   39999999.00  
pledged   4233.0  0.0   8596474.58    8596474.58  
backers     56.0  0.0     63416.00      63416.00  


> Deux derniers indicateurs nous permettent d'avoir un bon aperçu : **la variance** et **l'écart-type**. Il font partie de la famille des indicateurs de dispersion : 

* La variance mesure la dispersion autour de la moyenne, c'est à dire à quel point les valeurs présentes dans une liste numérique sont écartées de la moyenne. Elle se définit mathématiquement comme suit :


><center>   
$$ Var(X) = \dfrac{1}{N}\displaystyle\sum\limits_{i=1}^N(x_i-\overline{X})^2 $$
</center>

> Où $\overline{X}$ est la moyenne, N la longueur de la liste et $x_i$ les éléments de la liste.

* L'écart type, quant à lui, se définit mathématiquement comme la racine carrée de la variance. De par sa construction, **l’écart type** est également une mesure de la dispersion des valeurs. 

<div class="alert alert-info">
<i class="fa fa-info-circle"></i> &emsp; 
On peut calculer facilement l'écart type d'une variable en utilisant <code style="background-color:transparent;color:inherit">pandas.Series.std()</code>
</div>

* Calculer **l'écart type** de la variable **backers**.
* Afficher l'intervalle **moyenne** +/- **ecart type** de la variable **backers**.

In [39]:
# Insérez votre code ici





110.6665
intervalle: -850.1273671553156  ;  1071.4603671553157


<br>
<div class="alert alert-info">
<i class="fa fa-info-circle"></i> &emsp; 
On peut tout à fait tronquer l'intervalle à 0. Le nombre de backers est toujours positif.
</div>

* Ajouter une nouvelle colonne intitulée *backers_in*  à **`num_data`** contenant les valeurs de **backers** qui appartiennent à l'intervalle calculé précédemment.

<div class="alert alert-info">
<i class="fa fa-info-circle"></i> &emsp; 
Pour déterminer les valeurs qui appartiennent à l'intervalle calculé précédemment, on pourra s'aider de pandas.Series.between(int1,int2).astype(int).
</div>
<br>
<div class="alert alert-info">
<i class="fa fa-info-circle"></i> &emsp; 
Pour afficher les proportions on pourra s'aider de pandas.Series.value_counts(normalize = True).
</div>


In [37]:
# Insérez votre code ici






/home/ubuntu/virtualenvs/python3/lib/python3.6/site-packages/ipykernel_launcher.py:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  after removing the cwd from sys.path.


1    0.9856
0    0.0144
Name: backers_in, dtype: float64

> Nous venons de construire un intervalle qui contient la très grande majorité des données de **backers**. Il est utile pour déterminer facilement les valeurs extrêmes (outliers).

* Afficher les lignes de num_data qui ont une valeur 0 pour **backers_in** et les stocker dans un DataFrame *outliers*.
* Afficher la moyenne et la médiane de **backers** pour le jeu de données *outliers*.

In [38]:
# Insérez votre code ici






               goal     pledged  backers  backers_in
ID                                                  
1000235643  15000.0   124998.00     2784           0
1000428719  20000.0    47273.00     1261           0
1000725978  80000.0   843743.00     3784           0
1000776914  40000.0    47056.86     3080           0
1001265769  25000.0   227673.00     2278           0
...             ...         ...      ...         ...
1049522400  15000.0    36576.04     1402           0
1050044665   7500.0    54266.00     1506           0
1050045256  75000.0   377035.00     1773           0
1050138201  40000.0   695138.50    28818           0
1050139758  95000.0  2746655.75     2733           0

[144 rows x 4 columns]

Médiane :  1956.0 
Moyenne : 3833.090277777778


> La moyenne est très élevée, bien plus que la moyenne globale trouvée précédemment. Il en est de même pour la médiane. Ce procédé nous a permis d'identifier les valeurs extrêmes de **backers**.

* Exécuter cette cellule de code.

In [40]:
num_data['labels'] = pd.qcut(num_data['backers'], 4, labels = ['q1','q2','q3','q4'])

viz = pd.DataFrame(columns = ['q1','q2','q3','q4'], index = [0])

for j in num_data['labels'].unique():
    viz.loc[0,j] = num_data[num_data['labels']==j]['backers'].mean()
    
viz

/home/ubuntu/virtualenvs/python3/lib/python3.6/site-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.


,q1,q2,q3,q4
0,0.706873,6.27675,30.1028,407.606


> Nous avons affiché dans le DataFrame **viz**, la moyenne des valeurs de chaque quantile. On peut constater une importante différence entre q3 et q4. Cela signifie que le dernier quart des valeurs évolue sur des valeurs bien plus élevées que le reste. On a une vrai rupture entre q3 et q4.